include/swift/SIL/SILInstruction.h

```diff
@@ -2133,6 +2138,32 @@ class ApplyInstBase<Impl, Base, true>
   bool hasSemantics(StringRef semanticsString) const {
     return doesApplyCalleeHaveSemantics(getCallee(), semanticsString);
   }
+
+// SWIFT_ENABLE_TENSORFLOW
+private:
+  /// Predicate used to filter InoutArgumentRange.
+  struct OperandToInoutArgument {
+    const Impl &inst;
+    OperandToInoutArgument(const Impl &inst) : inst(inst) {}
+    Optional<SILValue> operator()(unsigned long i) const {
+      auto paramInfo = inst.getSubstCalleeConv().getParameters()[i];
+      auto argument = inst.getArgumentsWithoutIndirectResults()[i];
+      if (paramInfo.isIndirectMutating())
+        return argument;
+      return None;
+    }
+  };
+
+public:
+  using InoutArgumentRange =
+      OptionalTransformRange<IntRange<unsigned long>, OperandToInoutArgument>;
+  /// Returns all `@inout` and `@inout_aliasable` arguments passed to the
+  /// instruction.
+  InoutArgumentRange getInoutArguments() const {
```

Comment:
Note: `getInoutArguments` is currently defined in the "full application" case of `ApplyInstBase`. This required duplicating `asImpl` from the "partial application" `ApplyInstBase` superclass, which is a private function.

Other options:
- Define `getInoutArguments` as a top-level function in `Differentiation.cpp` specialized for just `ApplyInst`. No problems there.
- Move `getInoutArguments` to the "partial application" superclass `ApplyInstBase`. This requires more duplication because `getArgumentsWithoutIndirectResults` is defined only in the "full application" `ApplyInstBase`.